crates/ra_ide/src/hover.rs

```diff
@@ -335,6 +346,277 @@ fn hover_for_definition(db: &RootDatabase, def: Definition) -> Option<Markup> {
+            in_link = true;
+            link_target = Some(target.clone());
+            evt
+        }
+        Event::End(Tag::Link(link_type, _target, _)) => {
+            in_link = false;
+            Event::End(Tag::Link(link_type, link_target.take().unwrap(), CowStr::Borrowed("")))
+        }
+        Event::Text(s) if in_link => {
+            let (link_target_s, link_name) = callback(&link_target.take().unwrap(), &s);
+            link_target = Some(CowStr::Boxed(link_target_s.into()));
+            Event::Text(CowStr::Boxed(link_name.into()))
+        }
+        Event::Code(s) if in_link => {
+            let (link_target_s, link_name) = callback(&link_target.take().unwrap(), &s);
+            link_target = Some(CowStr::Boxed(link_target_s.into()));
+            Event::Code(CowStr::Boxed(link_name.into()))
+        }
+        _ => evt,
+    })
+}
+
+/// Rewrite documentation links in markdown to point to an online host (e.g. docs.rs)
+fn rewrite_links(db: &RootDatabase, markdown: &str, definition: &Definition) -> String {
+    let doc = Parser::new_with_broken_link_callback(
+        markdown,
+        Options::empty(),
+        Some(&|label, _| Some((/*url*/ label.to_string(), /*title*/ label.to_string()))),
+    );
+
+    let doc = map_links(doc, |target, title: &str| {
+        // This check is imperfect, there's some overlap between valid intra-doc links
+        // and valid URLs so we choose to be too eager to try to resolve what might be
+        // a URL.
+        if target.contains("://") {
+            (target.to_string(), title.to_string())
+        } else {
+            // Two posibilities:
+            // * path-based links: `../../module/struct.MyStruct.html`
+            // * module-based links (AKA intra-doc links): `super::super::module::MyStruct`
+            let resolved = try_resolve_intra(db, definition, title, &target).or_else(|| {
+                try_resolve_path(db, definition, &target).map(|target| (target, title.to_string()))
+            });
+
+            if let Some((target, title)) = resolved {
+                (target, title)
+            } else {
+                (target.to_string(), title.to_string())
+            }
+        }
+    });
+    let mut out = String::new();
+    cmark(doc, &mut out, None).ok();
+    out
+}
+
+#[derive(PartialEq, Eq, Hash, Copy, Clone, Debug)]
+enum Namespace {
+    Types,
+    Values,
+    Macros,
+}
+
+lazy_static!(
+    /// Map of namespaces to identifying prefixes and suffixes as defined by RFC1946.
+    static ref NS_MAP: HashMap<Namespace, (HashSet<&'static str>, HashSet<&'static str>)> = hashmap!{
+        Namespace::Types => (hashset!{"type", "struct", "enum", "mod", "trait", "union", "module"}, hashset!{}),
+        Namespace::Values => (hashset!{"value", "function", "fn", "method", "const", "static", "mod", "module"}, hashset!{"()"}),
+        Namespace::Macros => (hashset!{"macro"}, hashset!{"!"})
+    };
+);
```

Comment:
There's no need to use lazinless here. This is static data and can be stored s such:

```
static TYPES: &[&str] = ["type", "struct", "enum" ... ]
```